Add particle neighbor kernel method

jhossbach opened the pull request:
Closes #4620

Description of changes:
- add a short-range loop method to generate a list of the particle ids of all particles within interaction range of a central particle (to be used in machine-learned potentials based on descriptors of a particle's neighborhood)
- expose `cell_structure.max_range()` to the python interface for unit testing
- refactor short-range loop methods (dependency inversion and separation of concerns)

jngrad commented on src/core/cells.cpp:
```diff
@@ -160,6 +166,24 @@ std::vector<int> mpi_get_short_range_neighbors(int const pid,
                   mpi_get_short_range_neighbors_local, pid, distance, false);
 }
 
+std::vector<Particle *> get_particle_neighbors_local(Particle const &p) {
```
```suggestion
static auto get_particle_neighbors(Particle const &p) {
```

* `static`: this function doesn't need to be visible outside of this source file, so marking it as `static` hides it and allows the compiler to inline it
* `auto`: the return type can be deduced since there is only one `return` statement
* `_local`: no longer needed, we have removed almost all `mpi_*_local` functions at this point, developers should now assume that core functions run in parallel by default
* the corresponding prototype can be removed from `cells.hpp`

jngrad commented on src/core/cells.cpp:
```diff
@@ -160,6 +166,24 @@ std::vector<int> mpi_get_short_range_neighbors(int const pid,
+  std::vector<Particle *> ret;
+
+  auto const distance = *boost::min_element(cell_structure.max_range());
+  double const cutoff2 = distance * distance;
```
```suggestion
  auto const cutoff2 = Utils::sqr(distance);
```

jngrad commented on src/core/cells.cpp:
```diff
@@ -160,6 +166,24 @@ std::vector<int> mpi_get_short_range_neighbors(int const pid,
+  double const cutoff2 = distance * distance;
+  detail::search_neighbors_sanity_check(distance);
+
+  auto kernel = [&ret, &cutoff2](Particle const &p1, Particle const &p2,
```
```suggestion
  auto const kernel = [&ret, cutoff2](Particle const &p1, Particle const &p2,
```

jngrad replied:
As a side-note, you could consider the following construction:
```c++
  auto out_ids = std::back_inserter(ret);
  auto const kernel = [&out_ids, cutoff2](int i) {*out_ids = i;};
```
This make the intent clear that your lambda can only push new elements to the vector, and not modify existing elements, or clear all elements, or call the vector destructor.

jngrad commented on src/core/cells.cpp:
```diff
@@ -160,6 +166,24 @@ std::vector<int> mpi_get_short_range_neighbors(int const pid,
+
+  auto kernel = [&ret, &cutoff2](Particle const &p1, Particle const &p2,
+                                 auto vec) {
+    auto _p2 = const_cast<Particle *>(&p2);
```
This will most likely invoke undefined behavior. Why not use `std::vector<Particle const *>` storage instead? Note how the `const` is on the left of the asterisk to signify it's a mutable value type (requirement of `std::vector`) that points to a `const Particle` (requirement of the short-range loop).

jhossbach replied:
~The goal is to be able to change e.g. the force of neighboring particles with this method. I am removing the const qualifier from `short_range_neighbor_loop` instead.~
We can change it to be `std::vector<Particle const *>)`.

jngrad commented on src/core/cells.cpp:
```diff
@@ -160,6 +166,24 @@ std::vector<int> mpi_get_short_range_neighbors(int const pid,
+  detail::search_neighbors_sanity_check(distance);
+
+  auto kernel = [&ret, &cutoff2](Particle const &p1, Particle const &p2,
+                                 auto vec) {
```
```suggestion
                                 auto const &vec) {
```
The `auto` keyword cannot resolve to a reference type, even when you pass a reference-valued type to the function. As a result, your original kernel would call the vector copy constructor ([live example](https://godbolt.org/#z:OYLghAFBqd5QCxAYwPYBMCmBRdBLAF1QCcAaPECAMzwBtMA7AQwFtMQByARg9KtQYEAysib0QXACx8BBAKoBnTAAUAHpwAMvAFYTStJg1DIApACYAQuYukl9ZATwDKjdAGFUtAK4sGIM6SuADJ4DJgAcj4ARpjEEgCspAAOqAqETgwe3r7%2ByanpAiFhkSwxcVyJdpgOGUIETMQEWT5%2BAVU1AnUNBEUR0bEJtvWNzTltwz2hfaUDFQCUtqhexMjsHOYAzKHI3lgmG24AbtVExPvYJhoAgpc3ZltUWFQA1G4A8soAmgD67%2BFCABUAEpyNwAt5A25MLxEZ78VAQaGwtAMBQEZ4AegAVOYAGxYjFUObPEwAdis12eVOexEwBGWDDhADo0gAvTAQOb7Ck3UkAEVum0wtCUUJhqDhqARSIlKLRJLMuKJJPJt2pNLpDOZbI5XI2PLJAuuQoY%2BCoguuMueUQaEDR6BAIGODhI%2BzcoQI52ecvReOdxLJPPVtPpxEZ8Ig/u5gv5HAWtE48V4fg4WlIqE4bms1meCiWK0wCo2PFIBE0cYWAGsQPENPpOJJk%2BX05xeAoQHWy6m46Q4LAYIgUKgWEk6LFyJQ0COx3FgFwzAEsIc8KsAGp4TAAdzeSUYnBLNFoBFi7YgUWbUVCDQAnvveJfmMRr28otoTnfSFO2II3gxaLfu1ILAWEMYBxEA/BaRqY520AzBVGqGE1hLD1hWbWg8CiYgbw8LBmwIYg8BYO8FioAxgAUdctx3PduF4fhBBEMR2CkGRBEUFR1EA3QuH0UCUGzSx9Ew9tIAWVAkkcARYIAWgAdTEWhnnk%2BCCKYZT7X2PlTEsawzA0ZS3mAZS%2BXeL5fjef5gVBcEgTbYUTgyFxTVGPxeOCKYSjKPQUjSKTMk8FofPyfzei82ZbAcjoGC6EZApyXj2n82LJmKfpyiGbpXL0NFujC9KJAWPNllWQr6w4JNSBTNMMw4Z5VAADlxGTcUkZ5gGQZBnnnJkzGeCBcEIEgiy4OZeC7LQ5irGs6wTDhG1IYjayq5tarbDtS3LBY%2B0HKdR3oMgKAgPaZxAOcFyAzBlzXDdt13FMDzoY9iFPc9AIfG8Pw%2Bp8XzfBwPy/RgCF/f9m2A0DwLTSDHJg5t4MQ48P1Qua0wwrCcIwNY0wIoiSL4cjKNumiHvo2QmPEViGPkJQ1GbXQAgMIwBN0oS0dEzl00kjJZIU2glJU1Q1I0gh0C0nSrEsfTDOAez7H85z3HitzAlNfKZgy3yCgC7Jlc10LPIKxKouSiZsqNuXagmNXvMS02lZyq2DfVsrioLMq5sq6reFq%2BqmpatqOq6nq%2BoG/BThGsbNu7KbyoWpa6y9lsOHWzstum5a5o2JtALWqPJvjTgzGzmrWzzitSGOF6nMkIA%3D)) in the hot loop.

jngrad commented on src/core/cells.cpp:
```diff
@@ -192,6 +216,30 @@ std::vector<PairInfo> non_bonded_loop_trace() {
+std::vector<NeighborPIDs> get_neighbor_pids() {
+  std::vector<NeighborPIDs> ret;
+  auto pid_kernel = [&ret](Particle const &p,
+                           std::vector<Particle *> const neighbors) {
```
```suggestion
                           std::vector<Particle *> const &neighbors) {
```

jngrad replied:
General question: do you currently return a vector of pointers, because you would like to add more kernels in the future? If not, you could directly return a vector of `int` from the hot-loop kernel.

jhossbach replied:
Yes, the intent of `run_on_particle_neighbors` is to be able to find properties of neighboring particles, such as charge etc.

jngrad commented on src/core/cells.cpp:
```diff
@@ -192,6 +216,30 @@ std::vector<PairInfo> non_bonded_loop_trace() {
+  for (auto const &p : cell_structure.local_particles()) {
+    run_on_particle_neighbors(p, pid_kernel);
+  }
+  Utils::Mpi::gather_buffer(ret, comm_cart);
```
Maybe this gather operation belongs to the script interface. We should try to keep the core fully MPI-parallel and deal with MPI reductions in the interface. The interface has its own communicator, obtained via a call to `context()->get_comm()`. The `comm_cart` global variable used here will be removed in the near future.

jhossbach replied:
`Utils::Mpi::gather_buffer` does not take a `const boost::mpi::communicator &` which is what `context()->get_comm()` provides, it takes `boost::mpi::communicator`. Any reason why the `comm` should not be a const reference?

jngrad commented on src/core/cells.hpp:
```diff
@@ -119,6 +119,41 @@ boost::optional<std::vector<int>>
+/**
+ * @brief Get all neighbors of a particle.
+ *
+ * Uses cell_structure.max_range() to identify all neighbors of the particle.
```
Consider using Doxygen `@ref` with the proper class method namespace resolution to create a link. This way, if the method name or namespace changes, the CI will notify us this docstring has become outdated.

jngrad commented on src/core/cells.hpp:
```diff
@@ -119,6 +119,41 @@ boost::optional<std::vector<int>>
+ *
+ * Uses cell_structure.max_range() to identify all neighbors of the particle.
+ *
+ * @param p Particle const &
```
Avoid copy-pasting the argument type as description, since Doxygen already displays the argument type. Either write a description, or don't write `@param` if the argument is self-explanatory.

jngrad commented on src/core/cells.hpp:
```diff
@@ -119,6 +119,41 @@ boost::optional<std::vector<int>>
+ *
+ * @return vector of pointer of all neighboring particles.
+ */
+std::vector<Particle *> get_particle_neighbors_local(Particle const &p);
```
Can probably be removed, this is an implementation detail.

jngrad commented on src/core/cells.hpp:
```diff
@@ -119,6 +119,41 @@ boost::optional<std::vector<int>>
+ *		 std::vector<Particle *>)</tt>
+ */
+template <typename Kernel>
+void run_on_particle_neighbors(Particle const &p, Kernel kernel);
```
Remove this prototype and move the docstring to `cells.cpp`.

Function templates rarely need a prototype. It can actually be dangerous to have a prototype, because multiple translation units can write calls to instances of the function template, and the developer must guarantee that there is a translation unit where these template instances exist, and that this translation unit is linked against all translation units where the template instances are called.

jngrad commented on src/core/cells.cpp:
```diff
@@ -91,6 +91,12 @@ std::vector<std::pair<int, int>> get_pairs_filtered(double const distance,
+void serialize(Archive &ar, NeighborPIDs &n, const unsigned int /* version */) {
+  ar &n.pid;
+  ar &n.neighbor_pids;
+}
```
This code is never called. Consider running the test case in parallel with:
```diff
diff --git a/testsuite/python/CMakeLists.txt b/testsuite/python/CMakeLists.txt
index e878840e1..b4b408242 100644
--- a/testsuite/python/CMakeLists.txt
+++ b/testsuite/python/CMakeLists.txt
@@ -242,3 +242,3 @@ python_test(FILE observable_cylindricalLB.py MAX_NUM_PROC 2 LABELS gpu)
 python_test(FILE analyze_chains.py MAX_NUM_PROC 1)
-python_test(FILE analyze_distance.py MAX_NUM_PROC 1)
+python_test(FILE analyze_distance.py MAX_NUM_PROC 2)
 python_test(FILE analyze_acf.py MAX_NUM_PROC 1)
```

jngrad commented on src/python/espressomd/analyze.py:
```diff
@@ -192,6 +192,16 @@ class Analysis(ScriptInterfaceHelper):
 
+    particle_neighbor_pids()
+        Get a dictionary, where each key-value-pair represents a particle
+        and it's respective neighbors.
```
```suggestion
        Get a list of all short-range neighbors for each particle.
```
To avoid repetition with the return paragraph.

jngrad commented on src/script_interface/analysis/Analysis.cpp:
```diff
@@ -81,6 +82,15 @@ Variant Analysis::do_call_method(std::string const &name,
     return mpi_reduce_sum(context()->get_comm(), local);
   }
+  if (name == "particle_neighbor_pids") {
+    auto const neighbor_pids = get_neighbor_pids();
+    std::unordered_map<int, std::vector<int>> dict;
+    std::for_each(neighbor_pids.begin(), neighbor_pids.end(),
+                  [&dict](NeighborPIDs const &neighbor_pid) {
+                    dict[neighbor_pid.pid] = neighbor_pid.neighbor_pids;
+                  });
```
```suggestion
    std::unordered_map<int, std::vector<int>> dict;
    context()->parallel_try_catch([&]() {
      auto const neighbor_pids = get_neighbor_pids();
      std::for_each(neighbor_pids.begin(), neighbor_pids.end(),
                    [&dict](NeighborPIDs const &neighbor_pid) {
                      dict[neighbor_pid.pid] = neighbor_pid.neighbor_pids;
                    });
    });
```
Because the core function can throw an exception, we need a parallel try catch block.

jngrad reviewed (changes requested):
Overall, looks good to me. Please find below suggestions to improve performance and design.

jhossbach commented:
I have added your suggestions @jngrad, please re-review.
Some details:
 - Removed the prototype of `run_on_particle_neighbors` from `cells.hpp` and moved the docstring to the cpp file
 - Applied your suggestions to `get_particle_neighbors_local`
   - This method is not needed outside of cells.cpp, so I removed the .hpp declaration and moved the docstring to the definition in the .cpp file
 - Moved the `gather_buffer` operation to the script interface and added the `serialize` method for the custom `NeighborPIDS` struct to `script_interface/analysis/Analysis.cpp`
   - I had to change `gather_buffer` in `src/utils/include/utils/mpi/gather_buffer.hpp` to take a const reference of the comm

jngrad committed:
core: Remove unused MPI callbacks

jngrad committed:
core: Refactor short-range loop algorithms

Apply dependency inversion principle to short-range loop algorithms.
Let the script interface forward its MPI communicator to the core,
instead of relying on the deprecated ::comm_cart global variable.
Make all short-range loop consumers responsible for calling
on_observable_calc().

jhossbach committed:
Add particle neighbor kernel method

jngrad reviewed (approved):
Thanks!
